diff --git a/phpdotnet/phd/Package/PHP/ChunkedXHTML.php b/phpdotnet/phd/Package/PHP/ChunkedXHTML.php
index 0440afa..c23b621 100644
--- a/phpdotnet/phd/Package/PHP/ChunkedXHTML.php
+++ b/phpdotnet/phd/Package/PHP/ChunkedXHTML.php
@@ -46,14 +46,13 @@ HEADER;
             );
             $nextLink = "<li style=\"float: right;\"><a href=\"{$next["href"]}\">{$next["desc"]} »</a></li>";
         }
-        if ($parentId = Format::getParent($id)) {
+        $parentId = $id;
+        while ($parentId = Format::getParent($parentId)) {
             $up = array(
                 "href" => $this->getFilename($parentId) . $this->getExt(),
                 "desc" => $this->getShortDescription($parentId),
             );
-            if ($up['href'] != 'index.html') {
-                $upLink = "<li><a href=\"{$up["href"]}\">{$up["desc"]}</a></li>";
-            }
+            $upLink = "<li><a href=\"{$up["href"]}\">{$up["desc"]}</a></li>" . $upLink;
         }
 
         $nav = <<<NAV
@@ -67,7 +66,6 @@ HEADER;
 </div>
 <div id="breadcrumbs" class="clearfix">
   <ul class="breadcrumbs-container">
-    <li><a href="index.html">PHP Manual</a></li>
     {$upLink}
     <li>{$title}</li>
   </ul>
