diff --git a/phpdotnet/phd/Package/Generic/XHTML.php b/phpdotnet/phd/Package/Generic/XHTML.php
index ad9dc32..e6da8a5 100644
--- a/phpdotnet/phd/Package/Generic/XHTML.php
+++ b/phpdotnet/phd/Package/Generic/XHTML.php
@@ -679,7 +679,14 @@ abstract class Package_Generic_XHTML extends Format_Abstract_XHTML {
             $basename = basename($css);
             $dest = md5(substr($css, 0, -strlen($basename))) . '-' . $basename;
             if (@copy($css, $stylesDir . $dest)) {
-                $this->stylesheets[] = $dest;
+                // Change file name to include hash of the file contents instead of the file path
+                $oldPath = $stylesDir . $dest;
+                $dest = substr(md5_file($oldPath), 0, 8) . '-' . $basename;
+                if (rename($oldPath, $stylesDir . $dest)) {
+                    $this->stylesheets[] = $dest;
+                } else {
+                    trigger_error(vsprintf('Impossible to rename the %s file.', [$oldPath]), E_USER_WARNING);
+                }
             } else {
                 trigger_error(vsprintf('Impossible to copy the %s file.', [$css]), E_USER_WARNING);
             }
diff --git a/phpdotnet/phd/Package/PHP/CHM.php b/phpdotnet/phd/Package/PHP/CHM.php
index f172f26..fb9dd5e 100644
--- a/phpdotnet/phd/Package/PHP/CHM.php
+++ b/phpdotnet/phd/Package/PHP/CHM.php
@@ -15,176 +15,176 @@ class Package_PHP_CHM extends Package_PHP_ChunkedXHTML
     private $LANGUAGES = array(
         "hk"    => array(
             "langcode" => "0xc04 Hong Kong Cantonese",
-            "preferred_charset" => "CP950",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Big5",
             "preferred_font" => "MingLiu,10,0"
         ),
         "tw"    => array(
             "langcode" => "0x404 Traditional Chinese",
-            "preferred_charset" => "CP950",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Big5",
             "preferred_font" => "MingLiu,10,0"
         ),
         "cs"    => array(
             "langcode" => "0x405 Czech",
-            "preferred_charset" => "Windows-1250",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1250",
             "preferred_font" => self::DEFAULT_FONT,
         ),
         "da"    => array(
             "langcode" => "0x406 Danish",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT,
             "title" => "PHP Manualen"
         ),
         "de"    => array(
             "langcode" => "0x407 German (Germany)",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT,
             "title" => "PHP Handbuch",
         ),
         "el"    => array(
             "langcode" => "0x408 Greek",
-            "preferred_charset" => "Windows-1253",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1253",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "en"    => array(
             "langcode" => "0x809 English (United Kingdom)",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT,
             "title" => "PHP Manual",
         ),
         "es"    => array(
             "langcode" => "0xc0a Spanish (International Sort)",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "fa"    => array(
             "langcode" => "0x429 Persian",
-            "preferred_charset" => "Windows-1254",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1254",
             "preferred_font" => "Sylfaen,10,0",
         ),
         "fr"    => array(
             "langcode" => "0x40c French (France)",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT,
             "title" => "Manuel PHP"
         ),
         "fi"    => array(
             "langcode" => "0x40b Finnish",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "he"    => array(
             "langcode" => "0x40d Hebrew",
-            "preferred_charset" => "Windows-1255",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1255",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "hu"    => array(
             "langcode" => "0x40e Hungarian",
-            "preferred_charset" => "Windows-1250",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1250",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "it"    => array(
             "langcode" => "0x410 Italian (Italy)",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT,
             "title" => "Manuale PHP",
         ),
         "ja"    => array(
             "langcode" => "0x411 Japanese",
-            "preferred_charset" => "CP932",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "csWindows31J",
             "preferred_font" => "MS PGothic,10,0"
         ),
         "kr"    => array(
             "langcode" => "0x412 Korean",
-            "preferred_charset" => "CP949",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "EUC-KR",
             "preferred_font" => "Gulim,10,0"
         ),
         "nl"    => array(
             "langcode" => "0x413 Dutch (Netherlands)",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "no"    => array(
             "langcode" => "0x414 Norwegian (Bokmal)",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "pl"    => array(
             "langcode" => "0x415 Polish",
-            "preferred_charset" => "Windows-1250",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1250",
             "preferred_font" => self::DEFAULT_FONT,
             "title" => "Podręcznik PHP",
         ),
         "pt_BR" => array(
             "langcode" => "0x416 Portuguese (Brazil)",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT,
             "title" => "Manual do PHP",
         ),
         "ro"    => array(
             "langcode" => "0x418 Romanian",
-            "preferred_charset" => "ASCII//TRANSLIT//IGNORE",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1250",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "ru"    => array(
             "langcode" => "0x419 Russian",
-            "preferred_charset" => "Windows-1251",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1251",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "sk"    => array(
             "langcode" => "0x41b Slovak",
-            "preferred_charset" => "Windows-1250",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1250",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "sl"    => array(
             "langcode" => "0x424 Slovenian",
-            "preferred_charset" => "Windows-1250",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1250",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "sv"    => array(
             "langcode" => "0x41d Swedish",
-            "preferred_charset" => "Windows-1252",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1252",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "tr"    => array(
             "langcode" => "0x41f Turkish",
-            "preferred_charset" => "Windows-1254",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1254",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "uk"    => array(
             "langcode" => "0x422 Ukrainian",
-            "preferred_charset" => "Windows-1251",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "Windows-1251",
             "preferred_font" => self::DEFAULT_FONT
         ),
         "zh"    => array(
             "langcode" => "0x804 Simplified Chinese",
-            "preferred_charset" => "CP936",
+            "preferred_charset" => "UTF-8",
             "mime_charset_name" => "gb2312",
             "preferred_font" => "simsun,10,0"
         )
diff --git a/phpdotnet/phd/Package/PHP/ChunkedXHTML.php b/phpdotnet/phd/Package/PHP/ChunkedXHTML.php
index 0440afa..cc716b6 100644
--- a/phpdotnet/phd/Package/PHP/ChunkedXHTML.php
+++ b/phpdotnet/phd/Package/PHP/ChunkedXHTML.php
@@ -16,14 +16,17 @@ class Package_PHP_ChunkedXHTML extends Package_PHP_Web {
     }
 
     public function header($id) {
-        $title = Format::getLongDescription($id);
+        $sdesc = $this->indexes[$id]['sdesc'] ?? '';
+        $ldesc = $this->indexes[$id]['ldesc'] ?? '';
+        $title = ($sdesc && $ldesc && $sdesc !== $ldesc) ? "$sdesc — $ldesc" : Format::getShortDescription($id);
+        $onlinePage = 'https://www.php.net/manual/'.$this->config->language.'/'.$id.'.php';
         static $cssLinks = null;
         if ($cssLinks === null) {
             $cssLinks = $this->createCSSLinks();
         }
         $header = <<<HEADER
 <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
-<html>
+<html><!-- Online page at $onlinePage -->
  <head>
   <meta http-equiv="content-type" content="text/html; charset=UTF-8">
   <title>$title</title>
@@ -46,14 +49,13 @@ HEADER;
             );
             $nextLink = "<li style=\"float: right;\"><a href=\"{$next["href"]}\">{$next["desc"]} »</a></li>";
         }
-        if ($parentId = Format::getParent($id)) {
+        $parentId = $id;
+        while ($parentId = Format::getParent($parentId)) {
             $up = array(
                 "href" => $this->getFilename($parentId) . $this->getExt(),
                 "desc" => $this->getShortDescription($parentId),
             );
-            if ($up['href'] != 'index.html') {
-                $upLink = "<li><a href=\"{$up["href"]}\">{$up["desc"]}</a></li>";
-            }
+            $upLink = "<li><a href=\"{$up["href"]}\">{$up["desc"]}</a></li>" . $upLink;
         }
 
         $nav = <<<NAV
@@ -67,7 +69,6 @@ HEADER;
 </div>
 <div id="breadcrumbs" class="clearfix">
   <ul class="breadcrumbs-container">
-    <li><a href="index.html">PHP Manual</a></li>
     {$upLink}
     <li>{$title}</li>
   </ul>
diff --git a/phpdotnet/phd/Package/PHP/EnhancedCHM.php b/phpdotnet/phd/Package/PHP/EnhancedCHM.php
index 9be101c..9cbbcaf 100644
--- a/phpdotnet/phd/Package/PHP/EnhancedCHM.php
+++ b/phpdotnet/phd/Package/PHP/EnhancedCHM.php
@@ -74,7 +74,7 @@ class Package_PHP_EnhancedCHM extends Package_PHP_CHM
                 $this->outputHandler->v('Preparing usernotes.', VERBOSE_MESSAGES);
 
                 // Decompress the 'all' file into single files - one file per sectionid.
-                while($fpNotes && !feof($fpNotes) && false !== ($userNote = fgetcsv($fpNotes, 0, '|'))) {
+                while($fpNotes && !feof($fpNotes) && false !== ($userNote = fgetcsv($fpNotes, 0, '|', '"', '\\'))) {
                     // Usernote index
                     // 0 = Note ID
                     // 1 = Section ID
